File: claude_code_creators/app/javascript/channels/cloud_sync_channel.js
import consumer from "channels/consumer"

class CloudSyncChannel {
  constructor() {
    this.channel = null
    this.callbacks = new Map()
    this.currentUser = null
    this.integrationId = null
  }

  // Connect to the channel
  connect(userId, integrationId = null) {
    if (this.channel) {
      this.disconnect()
    }

    this.currentUser = userId
    this.integrationId = integrationId

    const params = { user_id: userId }
    if (integrationId) {
      params.integration_id = integrationId
    }

    this.channel = consumer.subscriptions.create(
      { channel: "CloudSyncChannel", ...params },
      {
        connected: () => {
          console.log("Connected to CloudSyncChannel")
          this.trigger("connected")
        },

        disconnected: () => {
          console.log("Disconnected from CloudSyncChannel")
          this.trigger("disconnected")
        },

        received: (data) => {
          console.log("Received cloud sync data:", data)
          this.handleMessage(data)
        }
      }
    )
  }

  // Disconnect from the channel
  disconnect() {
    if (this.channel) {
      consumer.subscriptions.remove(this.channel)
      this.channel = null
    }
  }

  // Handle incoming messages
  handleMessage(data) {
    const { event, integration_id, error } = data

    if (error) {
      this.trigger("error", { error, integration_id })
      return
    }

    switch (event) {
    case "sync_started":
      this.trigger("sync_started", data)
      break
    case "sync_progress":
      this.trigger("sync_progress", data)
      break
    case "sync_completed":
      this.trigger("sync_completed", data)
      break
    case "sync_error":
      this.trigger("sync_error", data)
      break
    case "file_imported":
      this.trigger("file_imported", data)
      break
    case "file_exported":
      this.trigger("file_exported", data)
      break
    default:
      // Handle status updates and other messages
      if (data.integrations) {
        this.trigger("status_update", data)
      } else if (integration_id) {
        this.trigger("integration_status", data)
      }
    }
  }

  // Request sync status
  getSyncStatus(integrationId = null) {
    if (!this.channel) return

    this.channel.perform("get_sync_status", {
      integration_id: integrationId
    })
  }

  // Trigger sync for an integration
  triggerSync(integrationId) {
    if (!this.channel) return

    this.channel.perform("trigger_sync", {
      integration_id: integrationId
    })
  }

  // Subscribe to events
  on(event, callback) {
    if (!this.callbacks.has(event)) {
      this.callbacks.set(event, [])
    }
    this.callbacks.get(event).push(callback)
  }

  // Unsubscribe from events
  off(event, callback) {
    if (this.callbacks.has(event)) {
      const callbacks = this.callbacks.get(event)
      const index = callbacks.indexOf(callback)
      if (index > -1) {
        callbacks.splice(index, 1)
      }
    }
  }

  // Trigger event callbacks
  trigger(event, data = {}) {
    if (this.callbacks.has(event)) {
      this.callbacks.get(event).forEach(callback => {
        try {
          callback(data)
        } catch (error) {
          console.error(`Error in CloudSyncChannel callback for ${event}:`, error)
        }
      })
    }
  }

  // Helper method to check if connected
  isConnected() {
    return this.channel && this.channel.consumer.connection.isOpen()
  }
}

// Create global instance
window.cloudSyncChannel = new CloudSyncChannel()

// Auto-connect if user data is available
document.addEventListener("DOMContentLoaded", () => {
  const userIdElement = document.querySelector("[data-user-id]")
  const integrationIdElement = document.querySelector("[data-integration-id]")
  
  if (userIdElement) {
    const userId = userIdElement.dataset.userId
    const integrationId = integrationIdElement?.dataset.integrationId
    
    window.cloudSyncChannel.connect(userId, integrationId)
  }
})

// Integrate with Stimulus controllers
document.addEventListener("cloud-sync:connect", (event) => {
  const { userId, integrationId } = event.detail
  window.cloudSyncChannel.connect(userId, integrationId)
})

document.addEventListener("cloud-sync:disconnect", () => {
  window.cloudSyncChannel.disconnect()
})

document.addEventListener("cloud-sync:get-status", (event) => {
  const { integrationId } = event.detail
  window.cloudSyncChannel.getSyncStatus(integrationId)
})

document.addEventListener("cloud-sync:trigger-sync", (event) => {
  const { integrationId } = event.detail
  window.cloudSyncChannel.triggerSync(integrationId)
})

export default window.cloudSyncChannel

File: claude_code_creators/app/channels/cloud_sync_channel.rb
class CloudSyncChannel < ApplicationCable::Channel
  def subscribed
    # Subscribe to cloud sync updates for the current user
    stream_for current_user

    # Also subscribe to specific integration channels if provided
    if params[:integration_id].present?
      integration = current_user.cloud_integrations.find_by(id: params[:integration_id])
      if integration
        stream_for integration
      end
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # Client can request sync status updates
  def get_sync_status(data)
    integration_id = data["integration_id"]

    if integration_id.present?
      integration = current_user.cloud_integrations.find_by(id: integration_id)
      if integration
        # Get current sync status (this would be enhanced with actual job tracking)
        status = {
          integration_id: integration.id,
          provider: integration.provider,
          syncing: false, # This would check actual sync job status
          files_count: integration.cloud_files.count,
          last_sync: integration.cloud_files.maximum(:last_synced_at),
          error: nil
        }

        transmit(status)
      end
    else
      # Return status for all user integrations
      statuses = current_user.cloud_integrations.map do |integration|
        {
          integration_id: integration.id,
          provider: integration.provider,
          syncing: false, # This would check actual sync job status
          files_count: integration.cloud_files.count,
          last_sync: integration.cloud_files.maximum(:last_synced_at),
          error: nil
        }
      end

      transmit({ integrations: statuses })
    end
  end

  # Client can trigger sync (with rate limiting)
  def trigger_sync(data)
    integration_id = data["integration_id"]

    return unless integration_id.present?

    integration = current_user.cloud_integrations.find_by(id: integration_id)
    return unless integration&.active?

    # Rate limiting: prevent sync requests more than once every 30 seconds
    cache_key = "sync_limit:#{current_user.id}:#{integration_id}"
    if Rails.cache.exist?(cache_key)
      transmit({
        error: "Sync rate limited. Please wait before syncing again.",
        integration_id: integration_id
      })
      return
    end

    # Set rate limit cache
    Rails.cache.write(cache_key, true, expires_in: 30.seconds)

    # Queue sync job
    CloudFileSyncJob.perform_later(integration)

    # Notify client that sync started
    transmit({
      event: "sync_started",
      integration_id: integration_id,
      provider: integration.provider,
      message: "Sync started for #{integration.provider_name}"
    })

    # Broadcast to all subscribers
    CloudSyncChannel.broadcast_to(current_user, {
      event: "sync_started",
      integration_id: integration_id,
      provider: integration.provider
    })
  end

  private

  def current_user
    # This assumes user is set in the connection
    connection.current_user
  end
end
